botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1075 @@
+         and matches the string value to an AWS partition.
+
+         :type value: str
+         :param value: A string representing an AWS region
+
+         :return: `output` object in the matched partition e.g.
+        {
+             "name": "aws",
+             "dnsSuffix": "amazonaws.com",
+             "dualStackDnsSuffix": "api.aws",
+             "supportsFIPS": true,
+             "supportsDualStack": true
+         }
+        """
+        if value is None:
+            return None
+
+        for partition in self.partitions['partitions']:
+            if (
+                value in partition['regions']
+                or re.match(partition['regionRegex'], value) is not None
+            ):
+                output = partition['outputs']
+                output['name'] = partition['id']
+                return output
+
+        return None
+
+    def _aws_parse_arn(self, value):
+        """**NOTE:** This is an internal API feature that is NOT
+        subject to backwards compatibility guarantees.
+
+        Evaluates a single string argument value, and returns an
+        object containing details about the parsed ARN.
+
+        :type value: str
+        :param value: A string representing an ARN
+
+        :return: a dict containing parsed components of the ARN string e.g.
+        {
+            "partition": "aws",
+            "service": "s3",
+            "region": "us-east-1",
+            "accountId": "012345678901",
+            "resourceId": ["user", "Development", "product_1234", "*"]
+        }
+        """
+        if value is None:
+            return None
+
+        if not value.startswith('arn:'):
+            return None
+
+        arn_parser = ArnParser()
+        try:
+            arn_dict = arn_parser.parse_arn(value)
+        except InvalidArnException:
+            return None
+
+        if any([len(component) == 0 for component in arn_dict.values()]):
+            return None
+
+        arn_dict['accountId'] = arn_dict.pop('account')
+
+        resource = arn_dict.pop('resource')
+        if ':' in resource:
+            resource_id = resource.split(':')
+        else:
+            resource_id = resource.split('/')
+        arn_dict['resourceId'] = resource_id
+
+        return arn_dict
+
+    def _string_equals(self, value1, value2):
+        """**NOTE:** This is an internal API feature that is NOT
+        subject to backwards compatibility guarantees.
+
+        Evaluates two string values value1 and value2 for
+        equality and returns a boolean if both values match.
+
+        :type value1: str
+        :param value1: a string to compare
+
+        :type value2: str
+        :param value2: a string to compare
+
+        :returns: a bool indicating if value1 and value2 equal each other
+        """
+        if not all(isinstance(val, str) for val in (value1, value2)):
+            raise EndpointInputParametersError(
+                msg='both values must be strings'
```

Comment:
I think we need a more descriptive error here or more metadata. If we surface this, how are we going to know what values and ruleset failed?

Reply:
hmmm that's fair. the trouble is this function isn't even aware of where these inputs come from and a rule isn't aware of the ruleset that its a part of (although that can easily be changed). The best simple solution I can think of is to raise an error here, catch it in `call_function` and re-raise with the information it has available?  I'm not entirely sure on this one.